magenta-lib/src/main/scala/magenta/deployment_type/CloudFormation.scala

```diff
@@ -44,44 +50,74 @@ object CloudFormation extends DeploymentType with CloudFormationDeploymentTypePa
         |when in conflict.""".stripMargin
   ).default(Map.empty)
+
   val createStackIfAbsent = Param[Boolean]("createStackIfAbsent",
     documentation = "If set to true then the cloudformation stack will be created if it doesn't already exist"
   ).default(true)
 
+  val secondsToWaitForChangeSetCreation = Param("secondsToWaitForChangeSetCreation",
+    "Number of seconds to wait for the change set to be created").default(15 * 60)
+
   val updateStack = Action("updateStack",
     """
       |Apply the specified template to a cloudformation stack. This action runs an asynchronous update task and then
       |runs another task that _tails_ the stack update events (as well as possible).
       |
     """.stripMargin
   ){ (pkg, resources, target) => {
-      implicit val keyRing = resources.assembleKeyring(target, pkg)
-      implicit val artifactClient = resources.artifactClient
-      val reporter = resources.reporter
-
-      val amiParameterMap: Map[CfnParam, TagCriteria] = getAmiParameterMap(pkg, target, reporter)
-
-      val cloudFormationStackLookupStrategy = getCloudFormationStackLookupStrategy(pkg, target, reporter)
-
-      val globalParams = templateParameters(pkg, target, reporter)
-      val stageParams = templateStageParameters(pkg, target, reporter).lift.apply(target.parameters.stage.name).getOrElse(Map())
-      val params = globalParams ++ stageParams
-      List(
-        UpdateCloudFormationTask(
-          target.region,
-          cloudFormationStackLookupStrategy,
-          S3Path(pkg.s3Package, templatePath(pkg, target, reporter)),
-          params,
-          amiParameterMap,
-          getLatestAmi(pkg, target, reporter, resources.lookup),
-          target.parameters.stage,
-          target.stack,
-          createStackIfAbsent(pkg, target, reporter),
-          true
-        ),
-        CheckUpdateEventsTask(target.region, cloudFormationStackLookupStrategy)
-      )
+    implicit val keyRing = resources.assembleKeyring(target, pkg)
+    implicit val artifactClient = resources.artifactClient
+    val reporter = resources.reporter
+
+    val amiParameterMap: Map[CfnParam, TagCriteria] = getAmiParameterMap(pkg, target, reporter)
+    val cloudFormationStackLookupStrategy = getCloudFormationStackLookupStrategy(pkg, target, reporter)
+
+    val globalParams = templateParameters(pkg, target, reporter)
+    val stageParams = templateStageParameters(pkg, target, reporter).lift.apply(target.parameters.stage.name).getOrElse(Map())
+
+    val userParams = globalParams ++ stageParams
+    val amiLookupFn = getLatestAmi(pkg, target, reporter, resources.lookup)
+
+    val stackTags = cloudFormationStackLookupStrategy match {
+      case LookupByTags(tags) => Some(tags)
+      case _ => None
     }
+
+    val rawChangeSetName = s"riff-raff-${target.parameters.build.projectName}-${target.parameters.stage}-${target.parameters.build.id}-${target.region.name}"
```

Comment:
Thinking more I think you may need to have:
 - `pkg.name` to distinguish between two different CFN blocks in the same riff-raff.yaml
 - `target.stack` to distinguish between stacks when deploying to multiple stacks (that could be in the same account)

I don't think you need region here as I presume it only needs to be unique in an account/region.

Reply:
As discussed IRL this smacks of being somewhat brittle (if another dimension is added to riff-raff for example). The Cloudformation deployment type already uses a millis wall clock to create a unique key for the template when the template is over the limit. Given that precedent we may as well use that approach here too and just suffer that it is impure.

Reminder that you can mess with joda time if that seems like a sane way of testing this: https://markhneedham.com/blog/2008/09/24/testing-with-joda-time/